node/lib/package/InstalledPackageSource.ts

```diff
@@ -0,0 +1,121 @@
+
+import { PackageInfo } from "./PackageInfo";
+import { PackageSource } from "./PackageSource";
+import { LocalPackageSource } from "./LocalPackageSource";
+
+import * as fs from "mz/fs";
+import * as path from "path";
+
+/**
+ * Reads package metadata and loads package information from a local directory
+ * where OpenT2T packages have been npm-installed.
+ */
+export class InstalledPackageSource extends LocalPackageSource {
+    /**
+     * Path to the directory where packages are cached. In order to enable requiring
+     * modules in these packages without using relative paths, the path should be a
+     * node_modules directory that is included in the search path.
+     */
+    public readonly cacheDirectory: string;
+
+    /**
+     * Creates a new InstalledPackageSource with a specified cache directory.
+     *
+     * @param {string} cacheDirectory  Path to the directory where packages are
+     *     npm-installed.
+     */
+    constructor(cacheDirectory: string) {
+        super(cacheDirectory);
+    }
+
+    /**
+     * Gets information about all OpenT2T packages currently available from the source.
+     * CAUTION: This is a potentially slow operation.
+     *
+     * @returns {Promise<PackageInfo[]>} An array of package information
+     * objects, or an empty array if the cache is empty
+     */
+    public async getAllPackageInfoAsync(): Promise<PackageInfo[]> {
+        let packageInfos: PackageInfo[] = [];
+        let directoryNames: string[] =
+                await LocalPackageSource.getSubdirectoriesAsync(this.cacheDirectory);
+
+        for (let i = 0; i < directoryNames.length; i++) {
+            let packageName: string = directoryNames[i];
+            if (packageName.startsWith("@")) {
+                let subdirectoryNames: string[] = await LocalPackageSource.getSubdirectoriesAsync(
+                        path.join(this.cacheDirectory, packageName));
```

Comment:
This code unavoidably makes lots of assumptions about the directory structure. My original design of having package.json specify module paths would have avoided making such assumptions.